### DISCUSSION TITLE
py_environment 'time_step' doesn't match 'time_step_spec' - I can't spot the difference

### DISCUSSION THREAD

**Discussion by ghost** [tensorflow/agents#751] [Q&A] [answered]
I'm trying to create a custom tf-agents environment for trading. When I try to validate it by calling utils.validate_py_environment(environment, episodes=1), I'm getting a ValueError 'time_step' doesn't match 'time_step_spec' . I've been trying to spot the difference for a while now but I can't seem to find it. Am I missing something?

Observation Action Spec

`self._action_spec = array_spec.BoundedArraySpec(
shape=(), dtype=np.int32, minimum=0, maximum=2, name='_action')

#representation of the enviroment: price + open position state
self._observation_spec = array_spec.BoundedArraySpec(
shape=(), dtype=np.float32, minimum=0, name='_observation')`

The Error Message

`---------------------------------------------------------------------------
ValueError Traceback (most recent call last)
in ()
1 env = TradingEnv(df=df.head(10), lkb=5,normalize=False,render=True)
----> 2 utils.validate_py_environment(env, episodes=1)

/usr/local/lib/python3.7/dist-packages/tf_agents/environments/utils.py in
validate_py_environment(environment, episodes, observation_and_action_constraint_splitter)
78 raise ValueError(
79 'Given time_step: %r does not match expected '
---> 80 'time_step_spec: %r' % (time_step, batched_time_step_spec))
81
82 action = random_policy.action(time_step).action

ValueError:

Given time_step: TimeStep(
{'discount': array(1., dtype=float32),
'observation': array([584.95, 582.3 , 581.7 , 582.6 , 582.9 , 584.65],
dtype=float32),
'reward': array(0., dtype=float32),
'step_type': array(0, dtype=int32)})

does not match expected

time_step_spec: TimeStep(
{'discount': BoundedArraySpec(shape=(), dtype=dtype('float32'),
name='discount', minimum=0.0,
maximum=1.0),
'observation': BoundedArraySpec(shape=(), dtype=dtype('float32'),
name='_observation',
minimum=0.0, maximum=3.4028234663852886e+38),
'reward': ArraySpec(shape=(), dtype=dtype('float32'), name='reward'),
'step_type': ArraySpec(shape=(), dtype=dtype('int32'),
name='step_type')})`

**Comment by sguada** [comment, marked as answer]
`observation': array([584.95, 582.3 , 581.7 , 582.6 , 582.9 , 584.65], dtype=float32),`
vs
`'observation': BoundedArraySpec(shape=(), dtype=dtype('float32'),`

The observation is a vector not just a scalar.

**Reply by ghost** [reply]
Thanks  Sergio
